### ISSUE TITLE
Check-in hangs/stuck/freezes at 0 countdown

### ISSUE THREAD

**Issue by cabal95** [SparkDevNetwork/Rock#2236]
### Prerequisites
* [X] Put an X between the brackets on this line if you have done all of the following:
    * Can you reproduce the problem on a fresh install or the [demo site](http://rock.rocksolidchurchdemo.com/)?
    * Did you include your Rock version number and [client culture](https://github.com/SparkDevNetwork/Rock/wiki/Environment-and-Diagnostics-Information) setting?
    * Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3ASparkDevNetwork+-repo%3ASparkDevNetwork%2FSlack) to see if your bug or enhancement is already reported?

### Description

Randomly our iPad based kiosks (and we actually saw it happen one time on a PC) get stuck at "0" on the waiting countdown screen. After some deep investigation with WireShark it appears to be caused by multiple postbacks.

From what we can tell, sometimes when the countdown hits zero only a single postback occurs (verified in WireShark). Other times dozens of postbacks happen within milliseconds of each other. The first few succeed and then the following ones seem to have their TCP sockets closed as soon as they open. It appears the close happens from the client side, but I'm not up enough on my TCP headers to know if the server is refusing those connections and the client is just cleaning up. Either way, the end result is that a (many) postback fails which seems to leave the page in unstable state.

We have never seen this happen on any other screen which makes me think it is related to the countdown script. There is a comment in the Welcome.ascx.cs file noting that this happens but special code was added to prevent it. I'm thinking this special code is sometimes not effective.

Primary logging this so it is tracked and there is a place to discuss it. Throwing a mention to @azturner as it appears he was the one to commit the original code (5 years ago, how good is your memory? ;) ).

### Versions

* **Rock Version:** 6.3 (probably all versions in the past 5 years)
* **Client Culture Setting:** en-US

**Comment by JimMichael** [comment]
I wonder if this is somehow related to performance of the client as well? Since 1. We've never seen it with our iPad Air2 units (no WIndows clients), and 2. Saw lots of performance/drag/slowdown issues (to the point of being unusable) on the iPad2 units we had prior to upgrading them all. Maybe a higher-perf client is masking the issue better.

**Comment by cabal95** [comment]
It is possible.  I did some research on that as well, it appears that UIWebView (which the app uses) has some serious memory leaks and it is a known issue to Apple that they have not cared about fixing it.  They have instead released a replacement control, but it only works on 10.9 and later. Once we fix this countdown issue I plan to see if the new replacement control fixes the core memory leak issue.

For reference, when I hooked up my iPad to the profiler I could see a massive memory leak on the countdown page. Like, 4-8 megabytes per minute easily, which adds up quick on a 1GB device. There was still a memory leak on the "check-in ready, tap Search" screen but it was much slower. Generally navigating from page to page would recover most of the memory, but not in all cases. It seems like the garbage collector in Javascript does not work properly inside a UIWebView. So the more Javascript activity there is the faster the memory leaks.

**Comment by cabal95** [comment]
Some more investigation, it appears that this is the series of events:
1) Check-in is loaded 10 minutes before check-in opens, initial countdown timer initialized.
2) Over the course of the 10 minutes, 59 partial postbacks happen to refresh the page causing 59 extra countdown timers to be initialized (the old ones do not display on screen anymore since their DOM element has been destroyed, but the onExpire function is still valid.
3) Countdown reaches 0 and all 60 onExpire timers fire.
4) the clearCountdown() function is called which tests if the $ActiveWhen.text() value, if not empty continue.
5) The clearCountdown() function blanks out the $ActiveWhen.text() value and initiates the refresh.
6) The Postback caused by the refresh rebuilds the DOM, which repopulates the $ActiveWhen.text() value because we fired about 400ms before the server recognized the check-in is open.
7) The next onExpire timer fires and begins at 4, causing the process to repeat since $ActiveWhen.text() has a value again.

**Comment by JimMichael** [comment]
Nice sleuthing!

**Comment by azturner** [comment]
So @cabal95, any suggestions on a fix for this? Could the refreshKiosk() function just be updated to set the $CountdownTimer to null?

**Comment by cabal95** [comment]
@azturner, I have a fix in testing at Shepherd.  Basically I am destroying the old $countdown object after postback but before the old DOM is destroyed. In testing it has worked fine (and also reduced some of the memory leaks on iPad).

Over the weekend we had two iPads get stuck at 0 one time (normally we were having 5-6 per service). It is possible those iPads just didn't pick up the new javascript code so we are going to monitor again this weekend and see if we have any more occurrences of stuck at 0.

**Comment by azturner** [comment]
@cabal95 I also just pushed some changes to that script as part of another change to add barcode-checkin as well. Noticed it was using `Page.ClientSript.RegisterStartupScript` instead of `ScriptManager.RegisterStartupScript`. Can you let me know if you seen any issues with those changes: https://github.com/SparkDevNetwork/Rock/commit/c38cd96a3bd7dbbbbcb676e3fa46f80180499b06#diff-3b0dc118c23bf465759da188bde59a19L137

**Comment by cabal95** [comment]
@azturner The primary change I made was to add this:
```
Sys.WebForms.PageRequestManager.getInstance().add_pageLoading(function () {{
        $('.countdown-timer').countdown('destroy');
    }});
```
The only possible issue I can see is I noticed that you removed the `Sys.Application.add_load` wrapper. So we would just have to test to make sure that the JS code doesn't execute before the new `add_pageLoading` code - though I don't think it would.

From the docs, `add_pageLoading` gets executed after a successful partial postback, but before the DOM is touched. So the countdown timer still exists and can be properly destroyed before it's re-created.
